Merge Mip 203

rdenarie opened the pull request:
This commit add the portlet Platform Name
This portlet display the branding name as a title

<!-- Ensure to provide github issue and task id in the title -->
<!-- Choose between feat and fix in the title to differenciate a new feature from a fix -->
<!-- Title format must be :
feat: FEATURE TITLE - MEED-XXXX - meeds-io/meeds#1234
or
fix: Fix TITLE - MEED-XXXX - meeds-io/meeds#1234
-->

<!-- Description : describe the feature/the fix by answering theses questions : -->
<!-- Why is this change needed?-->
<!-- Prior to this change, ...-->
<!-- How does it address the issue?-->
<!-- This change ...-->


<!-- Tips : 
Try To Limit Each Line to a Maximum Of 72 Characters
Provide links or keys to any relevant tickets, articles or other resources

Remember to
- Capitalize the subject line
- Use the imperative mood in the subject line
- Do not end the subject line with a period
- Separate subject from body with a blank line
- Use the body to explain what and why vs. how
- Can use multiple lines with "-" for bullet points in body
-->

rdenarie committed:
feat: Add portlet Platform name - meeds-io/MIPs#203 - MEED-9364

This commit add the portlet Platform Name
This portlet display the branding name as a title

rdenarie committed:
feat: Add portlet Platform Logo - meeds-io/MIPs#203 - MEED-9365

This commit add the portlet Platform Logo
This portlet display the branding Logo

rdenarie committed:
feat: Add portlet Login Sidebar - meeds-io/MIPs#203 - MEED-9366

This commit add the portlet Login Sidebar

rdenarie committed:
fix : Radius management is not applied on login portlet - meeds-io/MIPs#203 - MEED-9425

Before this fix, the radius from page layout is not applied to loginSidebar, platformName, and platformLogo portlet

This fix add v-app component so that radius are correctly applied.

rdenarie committed:
feat: Add portlet Login Form - meeds-io/MIPs#203 - MEED-9360

This commit add the portlet Login Form

rdenarie committed:
feat: Add extensions for portlet login form - meeds-io/MIPs#203 - MEED-9363

This commit add multiproviders and registration to login form

rdenarie committed:
feat: Apply layout compatibility to PlatformName portlet - meeds-io/MIPs#203 - MEED-9406

This commit update the PlatformName portlet to apply the layoyut default title.
With this, administrator can update the layout of this porlet with default layout options.

rdenarie committed:
feat: Add Platform Logo Settings - meeds-io/MIPs#203 - MEED-9420

This commit implements settings for Platform Logo Portlet

rdenarie committed:
feat: Add settings for sidebar Portlet - meeds-io/MIPs#203 - MEED-9367

This commit add settings for the sidebar portlet

rdenarie committed:
feat: Add settings for LoginForm Portlet - meeds-io/MIPs#203 - MEED-9368

This commit add settings for loginForm Portlet.

rdenarie committed:
feat: Add settings for external authentication providers in login form - meeds-io/MIPs#203 - MEED-9454

This commit add settings for external provider in login form portlet

rdenarie committed:
fix: Login Portlets (loginForm, sidebarLogin, platformName and platformLogo) cant be added twice on the same page - meeds-io/MIPs#203 - MEED-9515

Before this fix, the 4 login portlets cant be used twice or more on the same page
This problem is due to the portlet applicationId which is the same for all instances
This commit changes the appId to use the windowId into, so that, same portlet on same page have not hte same id.

rdenarie committed:
fix: Emoticon are not correctly displayed in loginForm translations - meeds-io/MIPs#203 - MEED-9520

Before this fix, translations in loginForm do not correctly display emoticon due to encoding problem
This commit ensure to pass correctly the encoded string so that emoticon are displayed

rdenarie committed:
feat: Add option to goback when user click on "Signin with email" - meeds-io/MIPs#203 - MEED-9513

Before this fix, when user clicks on "Signin with email" button, he cannot go back
This commit add a back button

rdenarie committed:
fix: When instanciating a new loginForm portlet, there is an error in server log - meeds-io/MIPs#203 - MEED-9521

Before this fix, when saving settings of a new instance of the portlet, in administration -> development -> portlet -> instances, there is an error in server log, related to unique constraint violation.
As saveTranslations request are sent in parallele, if there is more than one request, the service try to save the metadata more than once.

This commit ensure to make the first saveTranslation request synchronously, before the other request.
So when second and after request are launched, the metadata exists, and is not recreated.

rdenarie committed:
fix: Transform branding properties to portletPreferences for sidebarLogin - meeds-io/MIPs#203 - MEED-9487

Before this fix, siebarLogin portlet use Branding information for background image, alt text, title and subtitle
With this it is not possible to update the background for a portlet without changing for all portlets.

This commit transforms branding properties into portletPreferences

rdenarie committed:
fix: Adjust Login portlet to content - MEED-9470 - meeds-io/MIPs#203

Before this fix, the login form portlet take all the container width.
This commit limit the width to 375px

rdenarie committed:
fix: Apply primary color to the envelope icon - MEED-9650 - meeds-io/MIPs#203

Before this fix, the envelope icon is black instead of primary color

rdenarie committed:
fix: Go Back arrow is not aligned with text in login form - MEED-9548 - meeds-io/MIPs#203

Before this fix, the arrow to go back is not aligned with the text.
This commit modify margin to make it aligned.

rdenarie committed:
fix: Ability to not set text for login form - MEED-9649 - meeds-io/MIPs#203

This commit allow to hide the welcome message text in login form portlet

rdenarie committed:
fix: update display of button if long label, to use more than one line - MEED-9673 - meeds-io/MIPs#203

Add specific css to display button text on more than one line
This is not possible with standard configuration of vue component, we have to use css.

rdenarie committed:
feat: Create page login in global site - MEED-9372 - meeds-io/MIPs#203

Update lgin page component to make it work when the user is not connected
Before this fix, the translations for authentication providers cannot be translated because the user is anonymous and have not access to the rest service
This commit change the way translations are get to read it in jsp.

rdenarie committed:
fix: Missing settings in the login form drawer - MEED-9702 - meeds-io/MIPs#203

Before this fix, in some cases, the loginFormSettingsDrawer do not contains all settings.
This is because the providers are loaded before the drawer is created, so the js event is not registred in the drawer when it is launched.
This commit move the creation of the drawer before providers, so it ensure the drawer is correctly created before.

rdenarie committed:
feat : Make External Onboarding page modifiable in layout - MEED-9700 - meeds-io/MIPs#203

This commit update External Onboarding page to transform the existing handler and application into a portlet which can be added on a page, and configured with the layout editor.

rdenarie committed:
feat : Make Forgot password page modifiable in layout - MEED-9362 - Meeds-io/MIPs#203

This commit update Forgot password page to transform the existing handler and application into a portlet which can be added on a page, and configured with the layout editor.

rdenarie committed:
feat : Make Internal Onboarding page modifiable in layout - MEED-9711 - Meeds-io/MIPs#203
This commit update Internal Onboarding page to transform the existing handler and application into a portlet which can be added on a page, and configured with the layout editor.

rdenarie committed:
feat : Make Register page modifiable in layout - MEED-9361 - Meeds-io/MIPs#203

This commit update Register page to transform the existing handler and application into a portlet which can be added on a page, and configured with the layout editor.

rdenarie committed:
fix : Label for external providers is not updated once changed - MEED-9713 - meeds-io/MIPs#203

Before this fix, when updating provider translations, display is updated only after refresh
This commit ensure to propagate the event to the vue component so that the display is updating without refresh

rdenarie committed:
fix: Review of login page customization admin page - MEED-9375 - meeds-io/MIPs#203

This commit update the admin login page customization by replace all the branding part by links to access directly to default login pages

rdenarie committed:
fix : Unify UI of external providers with sign-in button - MEED-9763 - meeds-io/MIPs#203

External login providers buttons have different UI from other button on login page
This commit ensure to unify it.

rdenarie committed:
fix: Use css helper to set background-color for login buttons - MEED-9765 - meeds-io/MIPs#203

Before this fix, background color for login button are set with style directive
This modification use a css helper instead

rdenarie committed:
fix: Add label on back button in login form - MEED-9756 - meeds-io/MIPs#203

Before this fix, the back button have no label
This commit add a label for back button

rdenarie committed:
fix: use correct aria role for provider button - MEED-9757 - meeds-io/MIPs#203

Before this fix, a role="button" was assigned to provider button.
As it opens a new page,  it must not have button role but link role

This commit removes the v-tooltip component, which force the role=button for v-btn

rdenarie committed:
fix: On external login provider buttons, Remove accessible name when the text is displayed - MEED-9758 - meeds-io/MIPs#203

Before this fix, when the login provider button displays text, the button have accessible name, which is redundant.
This commit removes the accessible name when button already display it.

rdenarie committed:
fix: correct accesibility problems on login page - MEED-9202 - meeds-io/MIPs#203

Before this fix, the aria-label for hide/show password was not correct
In addition, the "signin" button have a not needed aria-label, as the text is already present on the button

rdenarie committed:
fix: remove tabindex on login form field - MEED-9786 - meeds-io/MIPs#203

This commit remove the uneeded tabindex field on login form text field

rdenarie committed:
fix: Add autocomplete attribute on login form fields - MEED-9638 - meeds-io/MIPs#203

This commit ensures to add the needed attributes autocomplete on fields of login form

rdenarie committed:
fix: Clean up code after feature completion - MEED-9753 - meeds-io/MIPs#203

Remove uneeded vue components

rdenarie committed:
fix: make settings drawers not loaded if user can not edit portlet preferences - MEED-9798 - meeds-io/MIPs#203

Before this fix, settings drawer for portlet sidebar login, login from, and platform logo were loaded in the page even if the user is not loggued and cannot edit
It generate unecessary load, and create accessibility error due to problems on component ExoDrawer

boubaker commented on component/api/src/main/java/io/meeds/social/translation/service/TranslationService.java:
```diff
@@ -32,6 +32,43 @@ public interface TranslationService {
+   *                                   object
+   * @throws ObjectNotFoundException When object wasn't found
+   */
+  Map<String, TranslationField> getAllTranslationFields(String objectType,
+                                                 String objectId,
+                                                 String username) throws IllegalAccessException, ObjectNotFoundException;
+
```
Code formatting

rdenarie replied:
done

boubaker commented on component/core/src/main/java/org/exoplatform/social/core/attachment/AttachmentServiceImpl.java:
```diff
@@ -360,6 +360,39 @@ public void moveAttachments(String sourceObjectType,
+                              String destinationObjectType,
+                              String destinationObjectId,
+                              String destinationParentObjectId,
+                              long userIdentityId, boolean deleteAfterMove) {
```
```suggestion
                              long userIdentityId,
                              boolean deleteAfterMove) {
```

rdenarie replied:
done

boubaker commented on component/core/src/test/java/io/meeds/social/translation/service/TranslationServiceTest.java:
```diff
@@ -39,7 +39,7 @@ public class TranslationServiceTest extends AbstractCoreTest { // NOSONAR
   private String                 objectType  = OBJECT_TYPE;
 
-  private long                   objectId    = 322l;
+  private String                   objectId    = "322";
```
```suggestion
  private String                 objectId    = "322";
```

rdenarie replied:
done

boubaker commented on webapp/src/main/java/io/meeds/social/portlet/CmsPortletWithMetadata.java:
```diff
@@ -0,0 +1,109 @@
+public class CmsPortletWithMetadata extends CMSPortlet {
+
+  public static final String OBJECT_TYPE = "cmsPortlet";
+  public static final String TRANSLATION_IDENTIFIER = "translationIdentifier";
```
```suggestion

  public static final String TRANSLATION_IDENTIFIER = "translationIdentifier";
```

rdenarie replied:
done

boubaker commented on webapp/src/main/java/io/meeds/social/portlet/CmsPortletWithMetadata.java:
```diff
@@ -0,0 +1,109 @@
+
+  private       TranslationService translationService;
+  private       AttachmentService  attachmentService;
+  private final SecureRandom       random = new SecureRandom();
```
```suggestion
  private       TranslationService translationService;

  private       AttachmentService  attachmentService;

  private  SecureRandomService secureRandomService;
```

rdenarie replied:
done

boubaker commented on webapp/src/main/java/io/meeds/social/portlet/CmsPortletWithMetadata.java:
```diff
@@ -0,0 +1,109 @@
+  @Override
+  protected void setViewRequestAttributes(String name, RenderRequest request, RenderResponse response) {
+    PortletPreferences preferences = request.getPreferences();
+    String currentTranslationIdentifier = preferences.getValue(TRANSLATION_IDENTIFIER,null);
```
```suggestion
    String currentTranslationIdentifier = preferences.getValue(TRANSLATION_IDENTIFIER, null);
```

rdenarie replied:
done

boubaker commented on webapp/src/main/java/io/meeds/social/portlet/CmsPortletWithMetadata.java:
```diff
@@ -0,0 +1,109 @@
+      // If translation identifier is not set, generate a new one
+      Long randomIdentifier = random.nextLong() & Long.MAX_VALUE;
+      currentTranslationIdentifier = String.valueOf(randomIdentifier);
+      savePreference(TRANSLATION_IDENTIFIER, currentTranslationIdentifier);
```
Why not using SETTING_NAME of the portlet ? Why creating a new Portlet Preference identifier ?

rdenarie replied:
I firstly tried with SETTING_NAME as for Image and it was not working.

I don't remember exactly, but there is a case, when you save a page template, or something like that, the setting_name change and we don't go in setViewRequestAttribute. It is not problematic for Image Portlet, but in case of translation, it make it not working as initial translations medatadas are not more associated with the correct identifier.

boubaker replied:
Page Template Management is different. Thus it should be better to fix the `SETTING_NAME` if there is a bug into it, rather than workaround it by redefining a new Portlet Preference IMO.

rdenarie replied:
What is the bug for you ? 
I'm not sure to understand

boubaker replied:
You said `but there is a case, when you save a page template, or something like that, the setting_name change and we don't go in setViewRequestAttribute`. If this behavior isn't ok and is a bug, thus we may need to fix it globally, isn't it ?

rdenarie replied:
For this point, I need to redig into current behaviour. 
Once done, I'll ping you to speak about that.

boubaker replied:
Sure :+1:

rdenarie replied:
I remember why I not use CMS_SETTINGS_NAME.

CMS_SETTINGS_NAME is on format "Untitled-603-d6964c9b-9480-4f22-8e56-b114e77b3e63"

But, as I save translations, I use translations services (which use metadata service), which require a long as ID for audience_id.

That's why I created a new settings, randomly generated

boubaker replied:
`audience_id` is a long yes which can be 0. No relation with the ObjectId = String.
In fact, there was a problem before which doesn't let using `TranslationService` with a String `ObjectId`, which has been improved already here: https://github.com/Meeds-io/social/commit/1f0ae4767460570e92478ff92d270a161b1f2aff

> feat: Improve TranslationService API with Object Id of Type String - [MEED-9632](https://builders.meeds.io/portal/meeds/tasks/taskDetail/9632) - https://github.com/Meeds-io/meeds/issues/3608

Thus I think we may now reuse the Setting Name. WDYT ?

rdenarie replied:
I had a version problem on my side.
It is working with the CMS SETTING NAME

Resolved

boubaker commented on webapp/src/main/webapp/vue-apps/common/components/ExoDrawer.vue:
```diff
@@ -417,6 +417,10 @@ export default {
         this.drawer = true;
       }
       this.resetFilter();
+      this.$nextTick(() => {
+        window.setTimeout(() =>
+          this.$el.querySelector('input')?.focus(), 50);
+      });
```
Why adding this generic behavior in all drawers ?

```suggestion
```

rdenarie replied:
For accessibility purpose : when you navigate with keyboard, and make an action opening a drawer, the focus is not "in" the drawer, so you cannot continue your navigation inside the drawer
This behavior ensure to put the focus on the first input of the drawer for all drawer of the platform.

In addition, it need to be improved in another task to take care of drawer which have no inputs.

boubaker replied:
Each drawer will have to manage this apart. In some drawers, not the first input is `autofocus-able`.

rdenarie replied:
Have you an example of a drawer in which the first input is not autofocus-able ?

This solution have the advantage to make all drawers a11y compatible in one modification

boubaker replied:
yes, drawers with `input = "checkbox"` or `input = "radio"` or `RichEditor` as first input. The `autofocus` attribute is added in all inputs Type (either it's of type `<input>` or not). The first assumption, that all inputs are of type `<input>` + the fact that it's available right after opening the Drawer is wrong.

Let's do a compromise, make this behavior optional using a dedicated attribute `autofocusFirstInput` with a `false` value by default. WDYT ?

rdenarie replied:
As discussed together, I'll add a paramter in exoDrawer, autofocusFirstInput, with default value = `true`

boubaker replied:
Yes, this may require some test on existing drawers (especially on mobile for Activity Stream Drawer). I think the focus will jump from the first input into the Rich Editor input. Can you check it please @srenault-meeds  ?

boubaker replied:
I thin even that it would be better to split concerns here and not couple the `Generic Components` accessibility with the current MIP. WDYT ? This way, we can take the time to study its impact.

srenault-meeds replied:
Hello, 

I agree to make the autofocus to first input. It is better for the UX actually. Indeed, it is painful for users to click again and again to finally access the input (for desktop and for mobile too).
And if this is good for accessibility, then let's keep it. Thanks.

Regarding the flag to make the autofocus activated or not, I confirm TRUE must be the default one. And then, it can be reviewed for some drawer in case it needs to be turned to FALSE.

Hope that helps

srenault-meeds replied:
FYI @boubaker we will remove this commit from the MIP and it will be done as a separated maintenance issue

boubaker replied:
ok, Thanks.

rdenarie replied:
Commit removed.

boubaker commented on webapp/src/main/webapp/vue-apps/login-form/components/LoginForm.vue:
```diff
@@ -0,0 +1,361 @@
+          :list-external-providers="listExternalProviders"
+          :display-providers-icons="displayProvidersIcons"
+          :display-welcome-message="displayWelcomeMessage"
+          v-if="values.canEdit"/>
```
Use `npm run lint` to format Vue files.

Besides, the `v-*` attributes goes all time in the first place
+ Drawers goes at the end of the file to let maintainers focus, easily, on Component content rather than Drawers (which are generally defined in a component apart)

rdenarie replied:
done

boubaker commented on webapp/src/main/webapp/vue-apps/login-form/components/LoginForm.vue:
```diff
@@ -0,0 +1,361 @@
+};
+</script>
+
+<style lang="scss">
+  .loginFormSigninEmailButton {
+    min-height: 36px;
+    height: 100% !important;
+  }
+  .loginFormSigninEmailButton > .v-btn__content {
+    max-width: 100%;
+    white-space: normal;
+  }
+</style>
```
```suggestion
```

Avoid defining inline Style inside the Vue files. This was banned from the beginning of Meeds.
The only usage where it was used, where the CSS is dynamically defined for animations. (Dynamic no Static)

rdenarie replied:
This css was added because Vue do not allow to display correctly a text on a button on 2 lines. 
I didnt manage to make the text wrap correctly without this trick.

Have you suggestions about that ?

boubaker replied:
yes, you can add the text inside a `div` or even a `<v-card max-width=".." width="..." color="transparent" flat>`:
```html
<v-btn...>
<div class="text-wrap">....</div>
</v-btn>
```

boubaker replied:
Thanks, seems fixed now.

rdenarie replied:
Not completly, I removed the css, but did not update the vue.
I have to finish this one

rdenarie replied:
I tried to replace the `v-btn` by a `v-card`. 
It is not a success. It creates a lot of code to reproduce the button behaviour, hover animation is still missing, it requires to have aria-pressed attributes in addition to role=button to be a11y compatible, and so on.

The current pb here, it that, by convention, Vue do not want to have button labels on 2 lines or more. They argue that button should stay short to preserve ergonomy. I'm completly ok with that point, but, in some language, like DE, short english word is a long german word. 

I'm also completly ok with the fact that we must ban specific css in vue file.

But, for me, it is a very specific case, and I think we should have an exception for LoginForm. I manage to reduce the css to only one instruction : I need to add max-width="100%" on 'v-btn__content'

boubaker replied:
you can add in `vuetify-all.less` ([here](https://github.com/Meeds-io/platform-ui/blob/develop/platform-ui-skin/src/main/webapp/skin/less/vuetify/vuetify-all.less#L280-L284)) any special behavior that you which to change on default `v-btn` coming from Vuetify. We already made some overrides there.

rdenarie replied:
are you sure that we should apply on all v-btn of the platform ?

boubaker replied:
by example:

```css
.v-btn.mutliple-lines {
  .v-btn__content {
    max-width: 100%;
  }
}
```

this way, you can add `class="mutliple-lines"` in your `v-btn` without changing the default behavior on other buttons.

rdenarie replied:
ok, deal. 
I do that.

Thanks

rdenarie replied:
Done

boubaker commented on webapp/src/main/webapp/vue-apps/login-form/components/LoginFormSettingsDrawer.vue:
```diff
@@ -0,0 +1,393 @@
+    :loading="loading"
+    eager
+    @closed="reset">
+    <template slot="title">
```
Avoid using `slot` since in Vue3, we will have to upgrade all to use `#` instead

```suggestion
    <template #title>
```

rdenarie replied:
done

boubaker commented on webapp/src/main/webapp/vue-apps/login-form/components/LoginFormSettingsDrawer.vue:
```diff
@@ -0,0 +1,393 @@
+
+      //if we have saveTranslations request, we need to ensure that the first one is executed before the others
+      //because it create the metadata for the translations
+      //if all saveTranslations are executed at the same time, we will have a "unique constraint violation" error because metadata will be created more than once
```
You can use `async` and `await` in this case which makes the method more readable.

rdenarie replied:
Can you detail what you mean ? 

In fact, I already use async and await when calling saveTranslationSynchronously to be sure that first saveTranslation request is done before doing the next promises.

boubaker replied:
It was just a proposition to make the code more clear and understandable in fact. But if not used to deal with `async` and `await`, then no problem, I 'll just resolve this remark and let you read more about it for next MIPs.

boubaker commented on webapp/src/main/webapp/WEB-INF/portlet.xml:
```diff
@@ -1131,4 +1131,183 @@
+    </portlet-info>
+  </portlet>
+
+
```
```suggestion
```

rdenarie replied:
done

rdenarie committed:
fix : Correct feedbacks for merge - meeds-io/MIPs#203 - MEED-9791

- Remove the usage of a specific preference translationIdentifier to use the cmsSettingName as identifier
- Fix code format
- use FormData to create url when fetching request
- Add parameter in PageLayoutRest and PageLayoutService to be able to update portletPreferences with the possibility to keep existing one
- Default behaviour is set to current behaviour : do no keep
- Update button Signin with Email in loginForm : remove the specific css to use a helper
- Finalize ForgotPassword transformation from handler to portlet
- Finalize Register transformation from handler to portlet
- Finalize External Onboarding transformation from handler to portlet
- Finalize Internal Onboarding transformation from handler to portlet
- Add Unit tests
- Reorganize fields in LoginRest
- Fix feedback in LoginService.js
- Fix feedback in ExternalOnboarding.vue
- Fix Test configuration
- Improve code lisibility
- Fix Providers loading in LoginFormSettingsDrawer : providers are loaded before the drawer, so the event in the drawer is never raised.
- Change a string "RegisterExtension" used few times to a constant
- Split methods in LoginRest to have one method for register, and one for verifyEmail

boubaker commented on webapp/src/main/webapp/vue-apps/login-form/main.js:
```diff
@@ -0,0 +1,45 @@
+import './initComponents.js';
+
+//getting language of the PLF
+const lang = typeof eXo !== 'undefined' ? eXo.env.portal.language : 'en';
```
```suggestion
const lang = eXo?.env?.portal?.language || 'en';
```

rdenarie replied:
done

boubaker commented on webapp/src/main/webapp/vue-apps/login-internal-onboarding/components/InternalOnboardingExpired.vue:
```diff
@@ -19,8 +19,8 @@
 <template>
-  <v-card flat>
-    <v-card-title class="primary--text text-font-size px-0">
+  <v-card flat class="transparent">
```
Attrs with attribute value goes before boolean attrs without value (Just a Hint)
+ you can use `color` on `v-card` instead of `class`

```suggestion
  <v-card color="transparent" flat>
```

rdenarie replied:
done

boubaker commented on webapp/src/main/webapp/vue-apps/login-internal-onboarding/components/InternalOnboardingResetForm.vue:
```diff
@@ -173,6 +177,33 @@ export default {
+      for (const [key, value] of new URLSearchParams(window.location.search)) {
+        url += `${key}=${value}&`;
+      }
+      const body = `action=resetPassword&username=${encodeURIComponent(this.username)}&password=${encodeURIComponent(this.password)}&password2=${encodeURIComponent(this.confirmPassword)}&captcha=${encodeURIComponent(this.captcha)}`;
```
Use `FormData.append`+ `new URLSearchParams(formData).toString()`, see it [here](https://github.com/Meeds-io/social/blob/develop/webapp/src/main/webapp/vue-apps/common/js/SpaceService.js#L193-L194)

rdenarie replied:
done for this file. Also applied to other concerned vue component

boubaker commented on webapp/src/main/webapp/vue-apps/login-internal-onboarding/main.js:
```diff
@@ -37,16 +35,16 @@ const urls = [
 ];
 
-export function init(params) {
+export function init(id,params) {
```
```suggestion
export function init(id, params) {
```

rdenarie replied:
done for this file, also applied for other vue components

boubaker commented on component/service/src/main/java/io/meeds/social/security/rest/LoginRest.java:
```diff
@@ -0,0 +1,712 @@
+public class LoginRest {
+
+  @Autowired
+  private PasswordRecoveryService passwordRecoveryService;
```
please format the file and use the following order for attributes definition:
1. Static
2. Public
3. Protected
4. Private

+ Attributes before Methods.

rdenarie replied:
done

boubaker commented on component/service/src/main/java/io/meeds/social/security/rest/LoginRest.java:
```diff
@@ -0,0 +1,712 @@
+                                             @Parameter(description = "Captcha name")
+                                             @RequestParam("name")
+                                             String name) {
+    HttpSession session = request.getSession();
```
Please ensure to split the Business Logic into Service Layer + Rest Layer. The Rest Layer should be kept as much simple as possible to access Contextual information (like Current User) and pass all into a Service Layer where we will have all the Business Logic

rdenarie replied:
This endpoint creates a random captcha image. 

I'm completely OK with the rule about no BusinessLogic in Rest Layer. 

But in this case, I think we should let this function in REST Layer : 

- Firstly, I estimate that there is no real BusinessLogic : no interaction with any other Services, no real condition to create the image or not.
- Secondly : if I create a Service Layer, it will contains only this function.

So on this case, I think I prefer to keep as is.

boubaker replied:
Yes, for the Captcha no problem (Even I prefer using an Utils Class rather than adding the source code in REST, in order to make it very light). It was a remark about all the `private` methods in the Bottom of the class, which holds some validation Business Logic:
- `findUser`
- `validateUserAndPassword`
- `validateUsernameAndPassword`
- `validateUserFullName`
- `validateEmail`
- `isExistsEmail`
- `sendOnboardingEmail`
- `generateUsername`
- `deleteFromInternalUsersGroup`
- `addToExternalUsersGroup`
- `generateUserDetailCredential`
- `finishUserCreation`
- `generateUserFromCredential`

All of those methods are business logic and should go to a service where we can reuse if we will do the same flow with other endpoints than REST: (Command Line, AI Tool, JSON RPC for MicroServices ...). The REST should get params + context info => pass them as is without thinking nor knowing what is it about + catch exception for response code specificity + format the output in HTTP response.

But no problem, let's keep it as is, we will have to rework some other REST Endpoints already but this one increases a bit the technical debt.

boubaker commented on webapp/src/main/java/io/meeds/social/portlet/CmsPortletWithMetadata.java:
```diff
@@ -0,0 +1,98 @@
+  protected void setViewRequestAttributes(String name, RenderRequest request, RenderResponse response) {
+    PortletPreferences preferences = request.getPreferences();
+    String initTranslationIdentifier = preferences.getValue(DATA_INIT_PREFERENCE_NAME,null);
+    if (initTranslationIdentifier != null) {
```
Why creating a new Translation if it's not an `export` operation ? IHO, we should keep it simple like made in other Portlets.

rdenarie replied:
To be able to change the translation on a page, without changing the translation of all instances. 

The translation is copied from the initial instance. But at the moment the portlet is added on a page, it have is own translation.

rdenarie replied:
When a portlet instance is create, a CMS Setting Name is generated. Translations done in this instance are associated with this CMS Settings Name

When the instance is added in a page, the LoginFormPortletInstancePreferencePlugin.generatePreferences returns the CMS Setting Name as DATA_INIT_PREFERENCE_NAME.
And then, the DATA_INIT_PREFERENCE_NAME is used to get back translations, and copy it to the new CMS Setting Name of the portlet.

This allow to get back existing translation if user add it in the portlet instance, and allow then update the new portlet on a page without modifying the initial instance.

boubaker replied:
Since the `DATA_INIT_PREFERENCE_NAME` will be used only the first time that a new portlet without `Setting Name` is added into a page. Thus I'm ok with the proposed first solution, until we work on the `export` mechanism.

boubaker commented on webapp/src/main/webapp/vue-apps/common/js/loginService.js:
```diff
@@ -0,0 +1,104 @@
+export function verifyToken(token, tokenType) {
+  let url = `/social/${eXo.env.portal.rest}/login/verify?token=${token}`;
+  if (tokenType) {
+    url = url.concat(`&tokenType=${tokenType}`);
```
User `FormData.append` to ensure that the URL is correctly encoded.

rdenarie replied:
done

boubaker commented on webapp/src/main/webapp/vue-apps/common/js/LoginService.js:
```diff
@@ -0,0 +1,104 @@
+export function verifyToken(token, tokenType) {
```
Please rename the file to start with Capital Letter

rdenarie replied:
done

boubaker commented on webapp/src/main/webapp/vue-apps/login-external-onboarding/components/ExternalOnboarding.vue:
```diff
@@ -19,27 +19,70 @@
+  }),
   created() {
-    document.title = this.$t('externalOnboarding.pagetitle');
+    this.token = new URLSearchParams(window.location.search).get('token');
```
```suggestion
    const searchParams = new URLSearchParams(window.location.search);
    this.token = searchParams.get('token');
    this.action = searchParams.get('action');
```

rdenarie replied:
done

boubaker commented on component/service/src/main/java/io/meeds/social/security/rest/LoginRest.java:
```diff
@@ -0,0 +1,763 @@
+      try {
+        String email = finishUserCreation(username, request, token, request.getLocale());
+        cleanCapcha(request.getSession(),"email-validation");
+        return ResponseEntity.status(HttpStatus.FOUND).header(LOCATION_HEADER, LOGIN_PATH + "?email="+email).build();
```
Can we avoid redirecting with user Email in HTTP URL (personal data inside the URL will be in clear text, even when using SSL)

rdenarie replied:
The idea is to prefill the login form with the user email.

Before modification, after email validation process, we auto-login the user. 
We the new architecture based on LoginRest, I'm not able to auto-login cleanely, and in addition, I'm not sure that it is a good security practice. 
So I change to redirect on login form, with identifier prefilled.

boubaker replied:
you can use the `sessionStorage` (not `localStorage`) to preserve the username/email to automatically login with in fact.

rdenarie replied:
sessionStorage is client side, right ?

I'm not sure to understand what you have in mean.
Can you explain ?

boubaker replied:
When the user registers and then the registration is ok (It's may be not using REST ?) then registration flow returns an error or a success. If error => display the error message right away without refreshing the page. If success => the entered email can be used to automatically send a user login Request using the entered login + password. WDYT ?
In client side, the business logic of navigation flow is controlled by JS Services and not in server side (which was the Stateful mode used before).

rdenarie replied:
to be able to finalize the merge, can we create this as other task and treat it after the merge ?
Thanks

boubaker replied:
I don't think that we can de-prioritize this since it will never be reworked after the merge. This is an unusual flow which may make the maintenance harder. IMHO, we need to make it clearer for maintainers, else this same code will have to be reworked and worked around each time it's unclear for a developer.

rdenarie replied:
I have few task to treat directly after the merge, before the 7.1 final release. 
So I can take this one during this set of task

That's why I propose to make it after the merge

boubaker replied:
ok, let's do it that way. Thanks.

boubaker commented on component/service/src/main/java/io/meeds/social/security/rest/LoginRest.java:
```diff
@@ -0,0 +1,763 @@
+      @ApiResponse(responseCode = "200", description = "Request fulfilled"),
+  })
+  public ResponseEntity<String> finishRegistration(HttpServletRequest request,
+                                            @Parameter(description = "Token")
```
Code format

rdenarie replied:
done

boubaker replied:
It appears not yet in fact.

```suggestion
  public ResponseEntity<String> finishRegistration(HttpServletRequest request,
                                                                                          @Parameter(description = "Token")
```

rdenarie replied:
Seems ok now on my side :
<img width="2201" height="463" alt="image" src="https://github.com/user-attachments/assets/0bdee626-ad66-4d09-91a3-59b215199c67" />

Can you recheck ?

boubaker replied:
ok, may be Github doesn't display well spaces, strange because the `verifyToken` is well displayed.
I mark this as resolved.

boubaker commented on component/service/src/main/java/io/meeds/social/security/rest/LoginRest.java:
```diff
@@ -0,0 +1,763 @@
+      if (!usernameToken.contains("@")) {
+        //username from token is not an email
+        //so we have not verify the email
+        //send accountVerificationEmail
```
We may need to explicitly split those two concerns into two dedicated methods to make it clearer, WDYT ?

rdenarie replied:
done

boubaker replied:
Sorry I always see an if => sendAccountVerificationEmail with else => createUser, switch whether the Token contains an `@` or not.

rdenarie replied:
Im sorry but I dont understand what you except.
In your préviens message you require to Split into 2 methods.
I created 2 methods : one for e-mail validation, one for direct création.

As in both case we are in a "register" use case, I prefer to keep only one register method in loginRest

boubaker replied:
We have to differentiate two different flows with two different methods. The outcome of `sendAccountVerificationEmail` is different from `createUser`. Thus, it's unclear why having both different flows in a single method. It's may clear for you, but unfortunately not clear enough for a developer willing to maintain the code. IMO, one REST method = One responsibility, the FrontEnd has to choose either the first method or the second one.

rdenarie replied:
Split done.
There are now 2 methods : 
- register
- VerifyEmail

boubaker replied:
Thanks :+1:

boubaker commented on component/service/src/main/java/io/meeds/social/security/rest/LoginRest.java:
```diff
@@ -0,0 +1,763 @@
+                                              @RequestParam("token")
+                                              String token) {
+
+    String usernameToken = passwordRecoveryService.verifyToken(token, CookieTokenService.FORGOT_PASSWORD_TOKEN);
```
Need to enhance this with an OTP code some day using `io.meeds.web.security.service.OtpService`. The code may be sent at the same time than the URL. Let's not change it for now, but will have to do it to secure the Reset Password mechanism.

boubaker replied:
Issue added here https://github.com/Meeds-io/meeds/issues/3734

boubaker commented on component/service/src/main/java/io/meeds/social/security/rest/LoginRest.java:
```diff
@@ -0,0 +1,763 @@
+                                            @RequestParam("captcha")
+                                            String captcha) {
+
+    String usernameToken = passwordRecoveryService.verifyToken(token, "onboard");
```
Same remark for OTP code flow usage, but will be treated apart in a dedicated enhancement.

boubaker replied:
Issue added here Meeds-io/meeds#3734

rdenarie commented:
> 

    To be able to change the translation on a page, without changing the translation of all instances.
    All instances are using the same CMS Setting Name ?

(I was not able to answer directly to your comment)

When a portlet instance is create, a CMS Setting Name is generated. Translations done in this instance are associated with this CMS Settings Name

When the instance is added in a page, the LoginFormPortletInstancePreferencePlugin.generatePreferences returns the CMS Setting Name as DATA_INIT_PREFERENCE_NAME.
And then, the DATA_INIT_PREFERENCE_NAME is used to get back translations, and copy it to the new CMS Setting Name of the portlet.

This allow to get back existing translation if user add it in the portlet instance, and allow then update the new portlet on a page without modifying the initial instance.

boubaker commented:
All points considered. Thanks.
